rlclientlib/utility/context_helper.cc

```diff
@@ -71,27 +61,110 @@ namespace reinforcement_learning { namespace utility {
     }
     catch ( ... ) {
-      RETURN_ERROR_LS(trace, status, json_parse_error) << error_code::unknown_s;
+      parse_error = error_code::unknown_s;
+    }
+
+    return count;
+  }
+
+  /**
+   * \brief Get the number of actions found in context json string.  Actions should be in an array
+   * called _multi under the root name space.  {_multi:[{"name1":"val1"},{"name1":"val1"}]}
+   *
+   * \param count   : Return value passed in as a reference.
+   * \param context : String with context json
+   * \param trace   : Pointer to the trace logger
+   * \param status  : Pointer to api_status object that contains an error code and error description in
+   *                  case of failure
+   * \return  error_code::success if there are no errors.  If there are errors then the error code is
+   *          returned.
+   */
+  int get_action_count(size_t& count, const char *context, i_trace* trace, api_status* status) {
+    std::string parse_error;
+    count = get_array_count(context, multi, trace, status, parse_error);
+    if (!parse_error.empty()) {
+      RETURN_ERROR_LS(trace, status, json_parse_error) << parse_error;
+    }
+
+    if (count > 0) {
+      return error_code::success;
     }
+
+    RETURN_ERROR_LS(trace, status, json_no_actions_found);
   }
 
+  /**
+   * \brief Get the number of slots found in context json string.  Slots should be in an array
+   * called _slots under the root name space.  {_slots:[{"name1":"val1"},{"name1":"val1"}]}
+   *
+   * \param count   : Return value passed in as a reference.
+   * \param context : String with context json
+   * \param trace   : Pointer to the trace logger
+   * \param status  : Pointer to api_status object that contains an error code and error description in
+   *                  case of failure
+   * \return  error_code::success if there are no errors.  If there are errors then the error code is
+   *          returned.
+   */
+  int get_slot_count(size_t& count, const char *context, i_trace* trace, api_status* status) {
+    std::string parse_error;
+    count = get_array_count(context, slots, trace, status, parse_error);
+    if (!parse_error.empty()) {
+      RETURN_ERROR_LS(trace, status, json_parse_error) << parse_error;
+    }
+
+    if ( count > 0 ) {
+      return error_code::success;
+    }
+
+    RETURN_ERROR_LS(trace, status, json_no_slots_found);
+  }
+
+  /**
+   * \brief Validate that the _multi and _slots entries exist in the context json string,
+   * and that _multi comes before _slots
+   * 
+   * \param context : String with context json
+   * \param trace   : Pointer to the trace logger
+   * \param status  : Pointer to api_status object that contains an error code and error description in
+   *                  case of failure
+   * \return  error_code::success if there are no errors.  If there are errors then the error code is
+   *          returned.
+   */
   int validate_multi_before_slots(const char *context, i_trace* trace, api_status* status)

```

Comment:
There's a comment in `live_model_impl.cc` that this was necessary due to limitations with the parser. Since this PR changes the parser, this function might not be necessary anymore. What exactly was the limitation?

Reply:
This is a limitation in VW itself and not the one on rlclientlib.